feat(mock-server): improve headers UI

Yogesh01000100 opened the pull request:
<!-- 🙌 Thanks for contributing to Requestly. Adding details below will help us to merge your PR faster. -->

Closes issue: #1134

## 📜 Summary of changes:

This pull request addresses the following changes related to header creation:

- [x] Implemented a dynamic way to add, edit, and remove headers.
- [x] Added error handling to display validation errors related to headers.
 
## Screen Replay session
https://app.requestly.io/sessions/saved/4diq8jHpNFIkFLebGCSE

https://github.com/requestly/requestly/assets/90953665/6743ca24-c584-4624-af56-bd140d7ae825






## ✅ Checklist:

- [x] Make sure linting and unit tests pass.
- [x] No install/build warnings introduced.
- [x] Verified UI in browser.
- [ ] For UI changes, added/updated analytics events (if applicable).
- [ ] For changes in extension's code, manually tested in Chrome and Firefox.
- [ ] For changes in extension's code, both MV2 and MV3 are covered.
- [ ] Added/updated unit tests for this change.
- [ ] Raised pull request to update corresponding documentation (if already exists).

## 🧪 Test instructions:

To test these changes, follow these steps:

1. Browse to the app folder and run the application.
2. Navigate to the Mock server from the sidebar section.
3. See for any already created mocks.
4. If it exists then click on it and check the response headers option.
5. Add new headers, edit existing ones, and remove headers.
6. Ensure that validation errors are displayed.

Yogesh01000100 commented:
@rohanmathur91 @RuntimeTerror10 @wrongsahil can you help merging this pull request?

wrongsahil commented:
@Yogesh01000100 Please attach [Session Replay](https://requestly.io/products/session-replays) in the PR description

Yogesh01000100 commented:
> @Yogesh01000100 Please attach [Session Replay](https://requestly.io/products/session-replays) in the PR description

please check the session, in the pull request description, (updated)

Yogesh01000100 commented:
> > @Yogesh01000100 Please attach [Session Replay](https://requestly.io/products/session-replays) in the PR description
> 
> @wrongsahil @RuntimeTerror10 please check the session, in the pull request description, (updated)

@RuntimeTerror10 can please check with the pull request that I put, and do let me know if any changes are required.

Yogesh01000100 commented:
@RuntimeTerror10 @wrongsahil can you check and help merge this PR?

RuntimeTerror10 commented:
Hey @Yogesh01000100 , I will review it today and will share any changes if required by tomorrow.

Yogesh01000100 commented:
> Hey @Yogesh01000100 , I will review it today and will share any changes if required by tomorrow.

okay sure!

Yogesh01000100 commented:
> Hey @Yogesh01000100 , I will review it today and will share any changes if required by tomorrow.

Have you reviewed it? if any changes are there please inform me.

RuntimeTerror10 commented on app/src/components/features/mocksV2/MockEditorIndex/Editor/HeaderSection/index.tsx:
```diff
@@ -0,0 +1,115 @@
+  const renderHeaderErrors = (errors: ValidationErrors, index: number, errorType: string) => {
+    if (errors.headers && errors.headers.some((err) => err.indexError === index)) {
+      return (
+        <div key={index}>
+          {errors.headers
+            .filter((err) => err.indexError === index)
+            .map((headerError, errorIndex) => {
+              if (headerError.valueError.startsWith(errorType)) {
+                const boxPosError = headerError.valueError.slice(errorType.length).trim();
+                return <div key={errorIndex}>{boxPosError}</div>;
+              }
+              return null;
+            })}
+        </div>
+      );
+    }
+    return null;
+  };
```
We should not perform checks on strings to determine if the error is for header `name` or `value`, error messages might change in future which will break this.

RuntimeTerror10 commented on app/src/components/features/mocksV2/MockEditorIndex/utils.ts:
```diff
@@ -48,3 +49,68 @@ export const getEditorLanguage = (fileType?: FileType) => {
+    const error: { typeOfError?: string; errorIndex?: number } = {};
+
+    if (!header.name) {
+      error.typeOfError = "name Header name is required";
```
Set type of error to either `name` or `value`, and add other property to error object named `description` which will be rendered below respective inputs.

RuntimeTerror10 replied:
Also remove the error type suffix added on each error message

RuntimeTerror10 commented:
@Yogesh01000100  Suggested UX change:
Instead of clicking on `Add modification` button to add a new header, can we automatically create a new empty header value-name pair when the user has added some value to the last header name-value field?

RuntimeTerror10 commented:
@Yogesh01000100 Added comments and one UX change

Yogesh01000100 commented:
> @Yogesh01000100 Added comments and one UX change

I've implemented the requested changes and added the new automatic header field creation upon user input.

Yogesh01000100 commented:
Hi @RuntimeTerror10 
It has been about two months since I submitted the pull request, and I haven't received any feedback or response. I understand that everyone is busy, and I appreciate your time and effort in reviewing pull requests.

I'm reaching out to inquire about the status of the review for this pull request. If there are specific concerns or changes needed, I'm more than willing to address them.

RuntimeTerror10 commented:
Hey @Yogesh01000100 
Sorry for the delay, I'll review your pull request and will get back to you soon with feedback.

Yogesh01000100 commented:
> Hey @Yogesh01000100 Sorry for the delay, I'll review your pull request and will get back to you soon with feedback.

 I understand that you're likely juggling a lot of responsibilities, and I appreciate the effort it takes to manage this project.

If there have been any developments regarding the review, or if there's any additional information or modifications you need from my end, please let me know. I'm eager to contribute and to make any necessary adjustments to my submission.

wrongsahil commented:
Closing this as of now as lot of things has changes in codebase.
